passthrough/usecase/passthrough_usecase.go

```diff
@@ -0,0 +1,95 @@
+package usecase
+
+import (
+	"context"
+	"strconv"
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	staking "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	"github.com/osmosis-labs/sqs/domain/mvc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+type passthroughUseCase struct {
+	poolsUseCase mvc.PoolsUsecase
+
+	bankQueryClient banktypes.QueryClient
+	stakingQueryClient staking.QueryClient
+}
+
+var _ mvc.PassthroughUsecase = &passthroughUseCase{}
+
+// NewPassThroughUsecase Creates a passthrough use case
+func NewPassThroughUsecase(grpcURI string, puc mvc.PoolsUsecase) (mvc.PassthroughUsecase, error){
+	grpcClient, err := grpc.Dial(grpcURI, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, err
+	}
+	
+	return &passthroughUseCase{
+		poolsUseCase: puc,
+
+		bankQueryClient: banktypes.NewQueryClient(grpcClient),
+		stakingQueryClient: staking.NewQueryClient(grpcClient),
+	}, nil
+}
+
+func (p *passthroughUseCase) GetAccountCoinsTotal(ctx context.Context, address string) (sdk.Coins, error) {
+ 	allBalancesRes, err := p.bankQueryClient.AllBalances(ctx, &banktypes.QueryAllBalancesRequest{
+		Address: address,
+	})
+	if err != nil {
+		return nil, err
+	}
+	
+	// Collect all balances from gamm shares and 
+	coins := sdk.NewCoins();
+	for _, balance := range allBalancesRes.Balances {
+		if strings.HasPrefix(balance.Denom, "gamm") {
+			// calc underlying coins from gamm shares
+			splitDenom := strings.Split(balance.Denom, "/")
+			poolID := splitDenom[len(splitDenom)-1]
+			poolIDInt, err := strconv.ParseInt(poolID, 10, 64)
+			if err != nil {
+				return nil, err
+			}
+
+			exitCoins, err := p.poolsUseCase.CalcExitCFMMPool(uint64(poolIDInt), balance.Amount)
+			if err != nil {
+				return nil, err
+			}
+			coins = coins.Add(exitCoins...)
+		} else {
+			coins = coins.Add(balance)
+		}
+	}
```

Comment:
There is a future optimization possible for this.

Currently, we query all balances and filter them out to contain gamm shares.

However, Cosmos SDK has a way to query KV store prefix directly.

As a result, we can query account KV store gamm balance prefix directly so that this would be an efficient database iteration.

I think this might save us quite a bit of performance.

Consider investigating this once the initial working MVP is ready but suggest not blocking on this comment at this stage.

Let me know if any links to helpful resources on this would help

Reply:
I'm thinking of supporting all the asset queries, then optimizing them as needed. Thoughts?

Reply:
SGTM